tests/Api.IntegrationTests/Behaviors/DeleteExpiredTokenKeysTests.cs

```diff
@@ -0,0 +1,60 @@
+using System.Net.Http.Json;
+using Bogus;
+using FluentAssertions;
+using Microsoft.Extensions.DependencyInjection;
+using Passwordless.Api.IntegrationTests.Helpers;
+using Passwordless.Service.Models;
+using Passwordless.Service.Storage.Ef;
+using Xunit;
+
+namespace Passwordless.Api.IntegrationTests.Behaviors;
+
+public class DeleteExpiredTokenKeysTests : IClassFixture<PasswordlessApiFactory>, IDisposable
+{
+    private readonly PasswordlessApiFactory _factory;
+    private readonly HttpClient _client;
+
+    public DeleteExpiredTokenKeysTests(PasswordlessApiFactory factory)
+    {
+        _factory = factory;
+        _client = factory.CreateClient().AddSecretKey();
+    }
+
+    [Fact]
+    public async Task An_expired_apps_token_keys_should_be_removed_when_a_request_is_made()
+    {
+        // Arrange
+        var tokenGenerator = new Faker<RegisterToken>()
+            .RuleFor(x => x.UserId, Guid.NewGuid().ToString())
+            .RuleFor(x => x.DisplayName, x => x.Person.FullName)
+            .RuleFor(x => x.Username, x => x.Person.Email)
+            .RuleFor(x => x.Attestation, "None")
+            .RuleFor(x => x.Discoverable, true)
+            .RuleFor(x => x.UserVerification, "Preferred")
+            .RuleFor(x => x.Aliases, x => new HashSet<string> { x.Person.FirstName })
+            .RuleFor(x => x.AliasHashing, false)
+            .RuleFor(x => x.ExpiresAt, DateTime.UtcNow.AddDays(1))
+            .RuleFor(x => x.TokenId, Guid.Empty);
+
+        using var scope = _factory.Services.CreateScope();
+
+        var storage = scope.ServiceProvider.GetRequiredService<ITenantStorageFactory>().Create("test");
+        await storage.AddTokenKey(new TokenKey { Tenant = "test", KeyMaterial = "randomKeyMaterial", KeyId = 0, CreatedAt = new DateTime(2022, 1, 1) });
+        await storage.AddTokenKey(new TokenKey { Tenant = "test", KeyMaterial = "randomKeyMaterial", KeyId = 1, CreatedAt = new DateTime(2022, 2, 1) });
+        await storage.AddTokenKey(new TokenKey { Tenant = "test", KeyMaterial = "randomKeyMaterial", KeyId = 2, CreatedAt = DateTime.UtcNow });
+
+        // Act
+        _ = await _client.PostAsJsonAsync("register/token", tokenGenerator.Generate());
+
+        // Assert
+        var tokenKeys = await storage.GetTokenKeys();
+        tokenKeys.Should().NotBeNull();
+        tokenKeys.Any(x => x.CreatedAt < (DateTime.UtcNow.AddDays(-30))).Should().BeFalse();
+        tokenKeys.Any(x => x.CreatedAt >= (DateTime.UtcNow.AddDays(-30))).Should().BeTrue();
+    }
```

Comment:
I would implement the test a bit differently:

1. Inject a fake `IDateTimeProvider` through the `WebApplicationFactory`
2. Simulate a behavior that would result in a new token being created (I lack the domain knowledge to know which one)
3. Move the time ahead in the future
4. Perform validations

In other words, we should not touch the database directly, but instead interact with the app through the means that are available to the user.

Also, I would not separate the test out like this, since it has to be tied out to some surface-level behavior (not necessarily endpoint) that's observable by the user, right?

Reply:
Good point. I'll make the change tomorrow morning...or later tonight if its really bugging me.

Reply:
> Also, I would not separate the test out like this, since it has to be tied out to some surface-level behavior (not necessarily endpoint) that's observable by the user, right?

Any endpoint that instantiates the `TokenServicce` and therefore the `Fido2ServiceEndpoints` creates this. Basically all the endpoints that return a token.  If I think about wholistic coverage, I'd have to replicate this test for each endpoint that returned a token...which seems like a lot of duplication.  Where would I put a test like this? Or would I simply do what I need to for the endpoint that is currently causing the error in production `signin/begin`?

Reply:
> Or would I simply do what I need to for the endpoint that is currently causing the error in production signin/begin?

I would do this. I assume most of the code you cover with this test is reused across other endpoints? In which case, you will not need to duplicate the test for each endpoint.

We have coverage maps now, so we can see where exactly we lack coverage – this way we don't have to approach testing like a numbers game but go off of what needs the most attention.

Reply:
In general, I think it's a decent approach to add tests only if we find a bug that could have been uncovered with that test. No need to overthink things too much.